### PR TITLE
Add FakeFirestore.FieldValue

### LINKED ISSUE #23: FakeFirestore doesn't declare FieldValue
# Description
I'm not quite sure what class is meant to mock over firebase.firestore (is that `FakeFirestore`?), but whatever it is doesn't seem to declare `FieldValue`.

## Steps to reproduce
1. Install and configure firestore-jest-mock and Jest.
```
npm install --save-dev jest firestore-jest-mock@latest
```
2. Have source code that uses `firebase.firestore.FieldValue`. In my case, I have a file that exports an object with a few Firebase helper types:
```
module.exports = {
  firestore: firebase.firestore(),
  Timestamp: firebase.firestore.Timestamp,
  TIMESTAMP: firebase.firestore.FieldValue.serverTimestamp(),
  increment: firebase.firestore.FieldValue.increment,
  arrayUnion: firebase.firestore.FieldValue.arrayUnion,
  arrayRemove: firebase.firestore.FieldValue.arrayRemove
};
```
3. Call `mockFirebase` in tests, before importing source files under test.
```
const { mockFirebase } = require("firestore-jest-mock");
mockFirebase({
  database: {
    ...
  }
});
```
4. Run `jest`

## Expected result
My tests should run, and fail for different reasons (failed assertions, etc).

## Actual result
No tests ran, and this error appeared in the console:
```
TypeError: Cannot read property 'serverTimestamp' of undefined

   7 |   firestore: firebase.firestore(),
   8 |   Timestamp: firebase.firestore.Timestamp,
>  9 |   TIMESTAMP: firebase.firestore.FieldValue.serverTimestamp(),
     |                                            ^
  10 |   increment: firebase.firestore.FieldValue.increment,
  11 |   arrayUnion: firebase.firestore.FieldValue.arrayUnion,
  12 |   arrayRemove: firebase.firestore.FieldValue.arrayRemove,
```

## Environment
- Node version: 8.17.0
- NPM version: 6.14.4
- Jest version: 25.5.3
- firebase-admin version: 8.11.0

### PR DESCRIPTION
**Pull request by AverageHelper**
# Description
Added a `firestore.FieldValue` mock to `FakeFirestore`. An instance of this new `FakeFirestore.FieldValue` class is normally created by calling one of its static functions: `arrayUnion`, `arrayRemove`, `increment`, `serverTimestamp`, and `delete`. These each should correspond to Firebase Firestore's own [`FieldValue` API](https://firebase.google.com/docs/reference/js/firebase.firestore.FieldValue).

Each `FakeFirestore.FieldValue` instance represents a sentinel that contains information on how to transform arbitrary values. If we so choose, we may modify `set`, `update`, and `add` to actively transform mock data using these sentinels.

The new `FakeFirestore.FieldValue` class can return transformed versions of arbitrary values by passing the value to its `transform` function.

## Related issues
Fixes #23 

## How to test
Run `jest` or `jest --verbose`.

### CODE REVIEW

**Comment by AverageHelper** [comment]
I'm still trying to get my fork to install right so I can test that things import correctly, but assuming I have the syntax down, this PR should do the trick.  Whaddya think?

**Comment by AverageHelper** [comment]
Ooh, found an issue. Don't merge this yet!

**Comment by AverageHelper** [comment]
Mmk, some issues here. Need to expose `firestore` with both static members and instance-like syntax, without having to prepend it with `new`...

**Comment by AverageHelper** [comment]
It worked!  It seems that named functions can have static members.  What a blessing!

**Comment by AverageHelper** [comment]
I'll keep playing with this, but now that my tests actually execute our firestore mock properly, I'm satisfied with what's here.  Feel free to clean it up or refactor it as you see fit.

**Comment by AverageHelper** [comment]
@sbatson5 I'm not sure who I'd ask about this, but have you or anyone else had a chance to look at this? Is there anything that needs fixed, or seems awkward to you guys?

I had an idea to scope the `mockX` functions such that `FakeFirestore.FieldValue.mockArrayRemove`, etc. would be a thing, rather than using the wordy `mockArrayRemoveFieldValue` name, but if y'all think this is reasonable and releasable in its current form, I'd be happy to see this PR closed.

**Comment by sbatson5** [comment]
Hi @AverageHelper, Sorry for the delay in getting back to you (work has been hectic for us). I will set some time aside this week to review 🙌

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "firestore-jest-mock",
-  "version": "0.3.0",
+  "version": "0.3.1",
```
Comment (sbatson5) [inline package.json:3]: Would you remind removing this change? We'll take care of the release version bump

Comment (AverageHelper) [reply package.json]: Done, and committed.

**Comment by AverageHelper** [comment]
Can do

**Comment by AverageHelper** [comment]
An idea that occurs to me just now is to move the new mock functions to static members of the relevant types.

For example, this PR adds the following as root-level exports:
```
const mockArrayRemoveFieldValue = jest.fn();
const mockArrayUnionFieldValue = jest.fn();
const mockDeleteFieldValue = jest.fn();
const mockIncrementFieldValue = jest.fn();
const mockServerTimestampFieldValue = jest.fn();
```
These are quite wordy, and to later add similar mock functions for Firestore's Transaction variant functions, this naming scheme would quickly become unwieldy.

What do you think about moving these new functions, and perhaps existing ones too (with aliases for compatibility), like so?
```
FakeFirestore.FieldValue.mockArrayRemove = jest.fn();
FakeFirestore.FieldValue.mockArrayUnion = jest.fn();
FakeFirestore.FieldValue.mockDelete = jest.fn();
FakeFirestore.FieldValue.mockIncrement = jest.fn();
FakeFirestore.FieldValue.mockServerTimestamp = jest.fn();
```
I'm fairly new to JavaScript classes, so I may not grasp the implications of this very well, but I think at least that these new method names express a little better where they're from without having to keep straight a slew of postfixes. Not only that, but each of these members would be exported simply by exporting `FakeFirestore` as we do.

Thoughts?

**Comment by AverageHelper** [comment]
My updates to README include the `FieldValue` suffixes on the new function mocks, in case that's the direction you wanna go with them.

**Comment by sbatson5** [comment]
@AverageHelper I agree that they are a bit wordy, but I'm ok with it because it's extremely clear and only adds wordiness to tests.

However, I don't think we can do `FakeFirestore.FieldValue.mockArrayRemove = jest.fn();` as that would declare a `jest.fn()` for each instance of a `FieldValue`, which would make it impossible to assert that they are called correctly.  By declaring the mocks outside of the context of the class, it allows us to look at that particular instance of the mock (since it's a singleton), which is what we want.

When mocking classes, in most cases, you want to do what you're describing.  However, since we are writing an abstracted layer and just want to assert that functions are called properly, we need to create `jest.fn()` outside the class.

Hope that makes sense!  What you have should work. Maybe we can move `FieldValue` to its own file to clean up some of the exports, but that can be a separate PR ;)
